File: python/calc24.py
#! /usr/bin/env python3

from enum import Enum
from dataclasses import dataclass
import random


class OpType(Enum):
    NONE = 0
    ADD = 1
    SUBSTRACT = 2
    MULTIPLY = 3
    DIVIDE = 4


@dataclass
class Operand:
    index: int
    op: OpType = OpType.NONE
    left: 'Operand' = None
    right: 'Operand' = None

    def to_string_helper(self, nums, is_denominator=False):
        if self.op == OpType.ADD or self.op == OpType.SUBSTRACT or (is_denominator and self.op != OpType.NONE):
            return f"({self.to_string(nums)})"
        else:
            return self.to_string(nums)

    def to_string(self, nums):
        if self.op == OpType.NONE:
            return str(nums[self.index])
        if self.op == OpType.ADD:
            return f"{self.left.to_string(nums)} + {self.right.to_string(nums)}"
        elif self.op == OpType.SUBSTRACT:
            return f"{self.left.to_string(nums)} - {self.right.to_string_helper(nums)}"
        elif self.op == OpType.MULTIPLY:
            return f"{self.left.to_string_helper(nums)} * {self.right.to_string_helper(nums)}"
        elif self.op == OpType.DIVIDE:
            return f"{self.left.to_string_helper(nums)} / {self.right.to_string_helper(nums, True)}"

    def eval(self, nums):
        if self.op == OpType.NONE:
            return float(nums[self.index])
        if self.op == OpType.ADD:
            return self.left.eval(nums) + self.right.eval(nums)
        elif self.op == OpType.SUBSTRACT:
            return self.left.eval(nums) - self.right.eval(nums)
        elif self.op == OpType.MULTIPLY:
            return self.left.eval(nums) * self.right.eval(nums)
        elif self.op == OpType.DIVIDE:
            return self.left.eval(nums) / self.right.eval(nums)


def combine(op1, op2):
    return [
        Operand(-1, OpType.ADD, op1, op2),
        Operand(-1, OpType.MULTIPLY, op1, op2),
        Operand(-1, OpType.SUBSTRACT, op1, op2),
        Operand(-1, OpType.SUBSTRACT, op2, op1),
        Operand(-1, OpType.DIVIDE, op1, op2),
        Operand(-1, OpType.DIVIDE, op2, op1),
    ]


def generate_formula(indexes):
    n = len(indexes)
    if n == 1:
        return indexes
    result = []
    for i in range(n):
        for j in range(i+1, n):
            reduced = [x for k, x in enumerate(indexes) if k != i and k != j]
            for r in combine(indexes[i], indexes[j]):
                reduced.append(r)
                result.extend(generate_formula(reduced))
                reduced.pop()
    return result


class Calc24:
    def __init__(self, n) -> None:
        indexes = [Operand(i) for i in range(n)]
        self.__formula = generate_formula(indexes)

    def calc(self, nums) -> str:
        target = float(24)
        for f in self.__formula:
            try:
                if f.eval(nums) == target:
                    return f.to_string(nums)
            except:
                continue
        return "No Solution"


if __name__ == '__main__':
    size = 4
    calc24 = Calc24(size)
    for _ in range(1000):
        nums = [random.randint(1, 13) for x in range(size)]
        result = calc24.calc(nums)
        print(f"{nums} -> {result}")

File: zig/calc24.zig
const std = @import("std");

const OpType = enum {
    None,
    Add,
    Subtract,
    Multiply,
    Divide,
};

const Operand = struct {
    index: usize = 0,
    op: OpType = .None,
    left: ?*Operand = null,
    right: ?*Operand = null,

    pub fn eval(self: @This(), nums: []u8) f64 {
        return switch (self.op) {
            .None => @floatFromInt(nums[self.index]),
            .Add => self.left.?.eval(nums) + self.right.?.eval(nums),
            .Subtract => self.left.?.eval(nums) - self.right.?.eval(nums),
            .Multiply => self.left.?.eval(nums) * self.right.?.eval(nums),
            .Divide => self.left.?.eval(nums) / self.right.?.eval(nums),
        };
    }

    pub fn toString(
        self: @This(),
        nums: []u8,
        arena: std.mem.Allocator,
    ) ![]const u8 {
        switch (self.op) {
            .None => return std.fmt.allocPrint(arena, "{d}", .{nums[self.index]}),
            .Add => {
                const l = try self.left.?.toString(nums, arena);
                const r = try self.right.?.toString(nums, arena);
                return std.fmt.allocPrint(arena, "{s} + {s}", .{ l, r });
            },
            .Subtract => {
                const l = try self.left.?.toString(nums, arena);
                const r = try self.right.?.toStringWrapped(nums, false, arena);
                return std.fmt.allocPrint(arena, "{s} - {s}", .{ l, r });
            },
            .Multiply => {
                const l = try self.left.?.toStringWrapped(nums, false, arena);
                const r = try self.right.?.toStringWrapped(nums, false, arena);
                return std.fmt.allocPrint(arena, "{s} * {s}", .{ l, r });
            },
            .Divide => {
                const l = try self.left.?.toStringWrapped(nums, false, arena);
                const r = try self.right.?.toStringWrapped(nums, true, arena);
                return std.fmt.allocPrint(arena, "{s} / {s}", .{ l, r });
            },
        }
    }

    pub fn toStringWrapped(
        self: @This(),
        nums: []u8,
        is_denominator: bool,
        arena: std.mem.Allocator,
    ) error{ OutOfMemory, AllocPrintError }![]const u8 {
        const r = try self.toString(nums, arena);
        if (self.op == .Add or self.op == .Subtract or (self.op != .None and is_denominator)) {
            return std.fmt.allocPrint(arena, "({s})", .{r});
        } else {
            return r;
        }
    }
};

fn combine(op1: *Operand, op2: *Operand, arena: std.mem.Allocator) ![6]*Operand {
    const add = try arena.create(Operand);
    add.* = .{ .op = .Add, .left = op1, .right = op2 };
    const mult = try arena.create(Operand);
    mult.* = .{ .op = .Multiply, .left = op1, .right = op2 };
    const sub1 = try arena.create(Operand);
    sub1.* = .{ .op = .Subtract, .left = op1, .right = op2 };
    const sub2 = try arena.create(Operand);
    sub2.* = .{ .op = .Subtract, .left = op2, .right = op1 };
    const div1 = try arena.create(Operand);
    div1.* = .{ .op = .Divide, .left = op1, .right = op2 };
    const div2 = try arena.create(Operand);
    div2.* = .{ .op = .Divide, .left = op2, .right = op1 };

    return [_]*Operand{ add, mult, sub1, sub2, div1, div2 };
}

fn generateFormula(indexes: std.ArrayList(*Operand), arena: std.mem.Allocator) !std.ArrayList(*Operand) {
    const n = indexes.items.len;
    var result = std.ArrayList(*Operand).init(arena);
    if (n == 1) {
        try result.appendSlice(indexes.items);
        return result;
    }
    var reduced = std.ArrayList(*Operand).init(arena);
    for (0..n) |i| {
        for ((i + 1)..n) |j| {
            reduced.clearRetainingCapacity();
            for (0..n) |k| {
                if (i != k and j != k) {
                    try reduced.append(indexes.items[k]);
                }
            }
            for (try combine(indexes.items[i], indexes.items[j], arena)) |r| {
                try reduced.append(r);
                const expanded = try generateFormula(reduced, arena);
                try result.appendSlice(expanded.items);
                _ = reduced.pop();
            }
        }
    }
    return result;
}

pub const Calc24 = struct {
    formula: std.ArrayList(*Operand),
    arena: std.mem.Allocator,

    pub fn init(comptime N: u8, arena: std.mem.Allocator) !Calc24 {
        var indexes = std.ArrayList(*Operand).init(arena);
        for (0..N) |i| {
            const x = try arena.create(Operand);
            x.* = .{ .index = i, .op = .None };
            try indexes.append(x);
        }
        return .{ .arena = arena, .formula = try generateFormula(indexes, arena) };
    }

    pub fn calc(self: @This(), nums: []u8) ![]const u8 {
        const target: f64 = 24;
        for (self.formula.items) |f| {
            if (f.eval(nums) == target) {
                return f.toString(nums, self.arena);
            }
        }
        return "No Solution";
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    const aa = arena.allocator();

    const stdout = std.io.getStdOut().writer();

    var seed: u64 = undefined;
    try std.posix.getrandom(std.mem.asBytes(&seed));
    var prng = std.rand.DefaultPrng.init(seed);
    const rand = prng.random();

    const N: i32 = 4;
    var nums: [N]u8 = undefined;
    var challenge = std.ArrayList(u8).init(aa);
    const calc = try Calc24.init(N, aa);
    for (0..100000) |_| {
        for (0..N) |i| {
            nums[i] = rand.intRangeAtMost(u8, 1, 13);
            if (i > 0) {
                try challenge.appendSlice(", ");
            }
            try challenge.writer().print("{d}", .{nums[i]});
        }
        const result = try calc.calc(&nums);
        try stdout.print("{s} -> {s}\n", .{ challenge.items, result });
        challenge.clearRetainingCapacity();
    }
}
